flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/GroupITCase.java

```diff
@@ -0,0 +1,807 @@
+		));
+	}
+
+	@Test
+	public void testGroupNextZeroOrMore() {
+		List<StreamRecord<Event>> inputEvents = new ArrayList<>();
+
+		Event c = new Event(40, "c", 1.0);
+		Event a1 = new Event(41, "a", 2.0);
+		Event b1 = new Event(42, "b", 3.0);
+		Event a2 = new Event(43, "a", 3.0);
+		Event b2 = new Event(44, "b", 3.0);
+		Event a3 = new Event(45, "a", 4.0);
+		Event b3 = new Event(46, "b", 3.0);
+		Event d = new Event(47, "d", 1.0);
+
+		inputEvents.add(new StreamRecord<>(c, 1L));
+		inputEvents.add(new StreamRecord<>(new Event(1, "event", 1.0), 2L));
+		inputEvents.add(new StreamRecord<>(a1, 3L));
+		inputEvents.add(new StreamRecord<>(b1, 4L));
+		inputEvents.add(new StreamRecord<>(a2, 5L));
+		inputEvents.add(new StreamRecord<>(b2, 6L));
+		inputEvents.add(new StreamRecord<>(a3, 7L));
+		inputEvents.add(new StreamRecord<>(b3, 8L));
+		inputEvents.add(new StreamRecord<>(d, 9L));
+
+		Pattern<Event, ?> pattern = Pattern.<Event>begin("start").where(new SimpleCondition<Event>() {
+			private static final long serialVersionUID = 6215754202506583964L;
+
+			@Override
+			public boolean filter(Event value) throws Exception {
+				return value.getName().equals("c");
+			}
+		}).next(Pattern.<Event>begin("middle1").where(new SimpleCondition<Event>() {
+			private static final long serialVersionUID = 5726188262756267490L;
+
+			@Override
+			public boolean filter(Event value) throws Exception {
+				return value.getName().equals("a");
+			}
+		}).followedBy("middle2").where(new SimpleCondition<Event>() {
+			private static final long serialVersionUID = 5726188262756267490L;
+
+			@Override
+			public boolean filter(Event value) throws Exception {
+				return value.getName().equals("b");
+			}
+		})).oneOrMore().optional().consecutive().followedBy("end").where(new SimpleCondition<Event>() {
+			private static final long serialVersionUID = 7056763917392056548L;
+
+			@Override
+			public boolean filter(Event value) throws Exception {
+				return value.getName().equals("d");
+			}
+		});
+
+		NFA<Event> nfa = NFACompiler.compile(pattern, Event.createTypeSerializer(), false);
+
+		final List<List<Event>> resultingPatterns = feedNFA(inputEvents, nfa);
+
+		compareMaps(resultingPatterns, Lists.<List<Event>>newArrayList(
+			Lists.newArrayList(c, d)
+		));
+	}
+
+	@Test
+	public void testGroupNest() {
+		List<StreamRecord<Event>> inputEvents = new ArrayList<>();
+
+		Event d = new Event(40, "d", 1.0);
+		Event a1 = new Event(41, "a", 2.0);
+		Event b1 = new Event(42, "b", 3.0);
+		Event c1 = new Event(43, "c", 4.0);
+		Event b2 = new Event(44, "b", 5.0);
+		Event c2 = new Event(45, "c", 4.0);
+		Event e = new Event(46, "e", 6.0);
+
+		inputEvents.add(new StreamRecord<>(d, 1));
+		inputEvents.add(new StreamRecord<>(a1, 2));
+		inputEvents.add(new StreamRecord<>(b1, 3));
+		inputEvents.add(new StreamRecord<>(c1, 4));
+		inputEvents.add(new StreamRecord<>(b2, 5));
+		inputEvents.add(new StreamRecord<>(c2, 6));
+		inputEvents.add(new StreamRecord<>(e, 7));
+
```

Comment:
Could you add as a comment  equivalents in pseudo pattern language? With the groups in work, it is extremely hard to read the patterns. Future readers would definitely benefit from it. 

I have in mind sth like : `d (a (b c)*)? e`
By default we would assume `followedBy` continuity, for others we would use `d any (a next(b c)*)? e` What do you think?